src/NuGet.Core/NuGet.ProjectManagement/Projects/MSBuildNuGetProject.cs

```diff
@@ -53,6 +53,16 @@ public class MSBuildNuGetProject : NuGetProject
         /// </summary>
         public event EventHandler<PackageEventArgs> PackageReferenceRemoved;
 
+        /// <summary>
+        /// Event to be raised when batch processing of install/ uninstall packages starts at a project level
+        /// </summary>
+        public event EventHandler<PackageEventArgs> BatchStart;
```

Comment:
These events and helpers should not be on the project. Having it on msbuild projects keeps it from working with project.json.

I would expect all of this to happen in the ExecuteNuGetProjectActionsAsync method. If you need to wrap preview also then additional tracking could be done within INuGetProjectContext